// @bcad1591/react-date-picker/dist/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWeekDay = exports.getYearMonth = exports.getThisYearAndThisMonth = exports.getPrevYearAndMonth = exports.getNextYearAndMonth = exports.getMonthData = exports.isBrowser = void 0;
var date_fns_1 = require("date-fns");
var isBrowser = function () {
    return typeof window !== 'undefined';
};
exports.isBrowser = isBrowser;
var getRangeArray = function (start, end) {
    return Array(end - start + 1)
        .fill(undefined)
        .map(function (_, index) { return index + start; });
};
var splitArray = function (array, part) {
    var tmp = [];
    for (var i = 0; i < array.length; i += part) {
        tmp.push(array.slice(i, i + part));
    }
    var lastTmpRow = tmp[tmp.length - 1];
    if (lastTmpRow.length < part) {
        var requiredCellCount = part - lastTmpRow.length;
        lastTmpRow.push.apply(lastTmpRow, Array(requiredCellCount).fill(false));
    }
    return tmp;
};
var getMonthData = function (year, month) {
    if (month < 1 || month > 12) {
        console.error('month must be between 1 and 12');
        return [];
    }
    var date = new Date(year, month - 1);
    var startDay = (0, date_fns_1.getDay)((0, date_fns_1.startOfMonth)(date));
    var endDate = (0, date_fns_1.getDate)((0, date_fns_1.endOfMonth)(date));
    var daysArray = getRangeArray(1, endDate);
    var spaceArray = Array(startDay).fill(false);
    var targetArray = spaceArray.concat(daysArray);
    return splitArray(targetArray, 7);
};
exports.getMonthData = getMonthData;
var getNextYearAndMonth = function (year, month) {
    var firstMonth = 1;
    var lastMonth = 12;
    if (month === lastMonth) {
        return [year + 1, firstMonth];
    }
    return [year, month + 1];
};
exports.getNextYearAndMonth = getNextYearAndMonth;
var getPrevYearAndMonth = function (year, month) {
    var firstMonth = 1;
    var lastMonth = 12;
    if (month === firstMonth) {
        return [year - 1, lastMonth];
    }
    return [year, month - 1];
};
exports.getPrevYearAndMonth = getPrevYearAndMonth;
var getThisYearAndThisMonth = function () {
    var today = new Date();
    return [(0, date_fns_1.getYear)(today), (0, date_fns_1.getMonth)(today)];
};
exports.getThisYearAndThisMonth = getThisYearAndThisMonth;
var getYearMonth = function (year, month) {
    var language = (0, exports.isBrowser)() ? navigator.language : 'en-US';
    return new Intl.DateTimeFormat(language, { year: 'numeric', month: 'long' }).format(new Date(year, month - 1));
};
exports.getYearMonth = getYearMonth;
var getWeekDay = function (locale) {
    if (locale.startsWith('ko')) {
        return ['일', '월', '화', '수', '목', '금', '토'];
    }
    if (locale.startsWith('ja')) {
        return ['日', '月', '火', '水', '木', '金', '土'];
    }
    if (locale.startsWith('zh')) {
        return ['日', '一', '二', '三', '四', '五', '六'];
    }
    return ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
};
exports.getWeekDay = getWeekDay;
