// @bcad1591/react-date-picker/dist/MonthTable.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YearMonthContext = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var styled_components_1 = __importDefault(require("styled-components"));
var Days_1 = __importDefault(require("./Days"));
var utils_1 = require("./utils");
var WeekDay_1 = __importDefault(require("./WeekDay"));
exports.YearMonthContext = (0, react_1.createContext)([0, 0]);
function MonthTable(_a) {
    var year = _a.year, month = _a.month;
    var days = (0, react_1.useMemo)(function () { return (0, utils_1.getMonthData)(year, month); }, [year, month]);
    var value = (0, react_1.useMemo)(function () { return [year, month]; }, [year, month]);
    return ((0, jsx_runtime_1.jsx)(exports.YearMonthContext.Provider, __assign({ value: value }, { children: (0, jsx_runtime_1.jsxs)(S.MonthTableLayer, { children: [(0, jsx_runtime_1.jsx)(S.Header, { children: (0, jsx_runtime_1.jsx)(S.YearMonth, { children: (0, utils_1.getYearMonth)(year, month) }) }), (0, jsx_runtime_1.jsxs)(S.Table, { children: [(0, jsx_runtime_1.jsx)(WeekDay_1.default, {}), (0, jsx_runtime_1.jsx)(Days_1.default, { days: days })] })] }) })));
}
var S = {
    YearMonth: styled_components_1.default.h2(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    font-size: 16px !important;\n    cursor: default !important;\n  "], ["\n    font-size: 16px !important;\n    cursor: default !important;\n  "]))),
    Header: styled_components_1.default.header(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    margin-bottom: 24px !important;\n    text-align: center !important;\n  "], ["\n    margin-bottom: 24px !important;\n    text-align: center !important;\n  "]))),
    Table: styled_components_1.default.table(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n    border-collapse: separate !important;\n    border-spacing: 0 2px !important;\n    font-weight: bold !important;\n  "], ["\n    border-collapse: separate !important;\n    border-spacing: 0 2px !important;\n    font-weight: bold !important;\n  "]))),
    MonthTableLayer: styled_components_1.default.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n    display: inline-block !important;\n    padding: 0 32px !important;\n  "], ["\n    display: inline-block !important;\n    padding: 0 32px !important;\n  "]))),
};
exports.default = (0, react_1.memo)(MonthTable);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
